
# SONAR 

## Configure

```{r,setup,warning = FALSE, message = FALSE}
library(SONAR)
library(here)
library(Matrix)
library(data.table)
library(Seurat)
library(matlabr)
library(R.matlab)
```
## Set paths / Prepare the Input

```{r}
#Set paths.In this example, we set the path as this files' dir.
knitr::opts_knit$set(root.dir = here::here())
options (warn = -1)
```
### #scRNA-seq data
```{r,warning=FALSE}
#you can prepare your data and replace these files with yours,please keep the same format

##The path to scRNA data
sc_path <- file.path(here::here("inst", "extdata", "sc"))

##Expression matrix(reffile) with rowname(gene) and colname(cell barcode)
reffile <- "development_heart.scRNA.processed.cnt.genexrow.tsv"

##Cell type annotation(metafile), which has 2 columns, the first col is cell barcode(same as colname in reffile), the second is cell type annotation. This file has header of colnames:"cellname","celltype"
metafile <- "type_annotation.tsv"
```
### #Spatial transcriptomics data
```{r,warning=FALSE}
#you can prepare your data and replace these files with yours,please keep the same format

#The path to spatial data
st_path <- file.path(here::here("inst", "extdata", "spatial"))

##Spatial expression numerical matrix(stfile) with rowname(gene) and colname(spot barcodes)
stfile <- "PCW6.5_1_st_cnt.genexrow.tsv"

##Spot coordinate(locfile), which has 3 cols, the first is spot barcodes(same as colname in stfile), the second is x coord(numeric), the third is y coord(numeric). This file has header of colnames: "barcodes","xcoord","ycoord"
locfile <- "coords.tsv"

###path of core-code, which doesn't need to modify
code_path <- file.path(here::here("core-code"), "/")
```

## Set the Output path
```{r,warning=FALSE}
#This is the path to output
result_path <- file.path(here::here("result"), "/")
```
## Pre-process and Deliver data to SONAR

```{r,message=FALSE, warning=FALSE}
#read in scRNA data
ref <- fread(file.path(sc_path, reffile))
ref <- as.data.frame(ref, check.names = FALSE)
rownames(ref) <- ref[,1]
ref[,1] <- NULL

cluster <- fread(file.path(sc_path, metafile))
cluster <- as.data.frame(cluster, check.names = FALSE)
typeanno <- cluster$celltype
names(typeanno) <- cluster$cellname
typeanno <- as.factor(typeanno)

#read in spatial data
spots <- fread(file.path(st_path, stfile))
spots <- as.data.frame(spots, check.names = FALSE)
rownames(spots) <- spots[,1]
spots[,1] <- NULL
spots <- spots[rowSums(spots) > 0,]

coords <- fread(file.path(st_path, locfile))
coords <- as.data.frame(coords, check.names = FALSE)
rownames(coords) <- coords[,1]
coords[,1] <- NULL
colnames(coords) <- c("x","y")

#get the overlap genes
overlap_gene <- intersect(rownames(spots), rownames(ref))
ref <- ref[overlap_gene,]
spots <- spots[overlap_gene,]

#calculate the nUMI and nUMI_spot
nUMI <- colSums(ref)
names(nUMI) <- colnames(ref)
nUMI_spot <- colSums(spots)
names(nUMI_spot) <- colnames(spots)

#preprocess the input data
processed_input<-SONAR.preprocess(sc_count=ref,sc_cell_type=typeanno,sc_nUMI=nUMI,sp_coords=coords,sp_count=spots,sp_nUMI=nUMI_spot,cores=8)

#deliver the preprocessed data to SONAR
SONAR.deliver(processed_data=processed_input,path=code_path)

#define the bandwidth, default is 1.2 times minimal distance
temp<-dist(coords)
temp<-Matrix::Matrix(temp)
temp[temp==0] <- NA
mindist <- min(temp,na.rm = T)
h <- 1.2*mindist
```
## Start Deconvolution and Collect results
```{r,warning=FALSE}
#start deconvolution
SONAR.deconvolute(fname = paste0(code_path,"SONAR_main.m"),path=code_path,h,wait = TRUE)
## if return 0, deconvolution has done

#collect results
SONAR.results <- readMat(paste(code_path,"SONAR_results.mat",sep = ""))
SONAR.results <- SONAR.results$JIE
u <- fread(paste0(code_path,"u.txt"))
u[,1] <- NULL
colnames(SONAR.results) <- colnames(u)
spot_name <- read.table(file=paste0(code_path,"coord.txt"),sep=",")
rownames(SONAR.results) <- rownames(spot_name)

#Complete! SONAR.results is the final results
write.table(SONAR.results,file = paste(result_path,'SONAR.results.txt',sep=""),sep = ",")

# #For Slide-seqV2 or datasets with higher resolution, we recommand to deal with the result to keep the cell types with top n (default n =3) high predictions in each spot
# topn <- function(x) {
#   z <- rep(0, length(x))
#   #if you want to keep less cell type, you can use 1:n, n is the cell type numbers that you want to keep
#   i <- order(x, decreasing = TRUE)[1:3]
#   z[i] <- x[i]
#   # return the vector
#   return(z)
# }
#
# # apply the function to each row of df
# SONAR.results <- apply(SONAR.results, 1, top2)
# 
# # convert back to data frame
# SONAR.results <- as.data.frame(t(SONAR.results))
# SONAR.results<-SONAR.results/rowSums(SONAR.results)

```

## Visualization

### #Pie plot to show the predicted cell type composition

```{r,warning=FALSE,fig.height=20,fig.width=22}
library(gtools)
library(ggplot2)
library(scatterpie)
library(viridis)
library(corrplot)
library(grDevices)
library(ggcorrplot)
#prepare the color/coords/results(SONAR.results)
color_vector<-c("#8DD3C7","#C7E2E1","#FB8072","#FFED6F","#80B1D3","#FDB462","#B3DE69","#FCCDE5","#D9D9D9","#BDB8D5","#CCEBC5","#D1C3D1","#A6CEE3","#1F78B4","PaleTurquoise","#6A3D9A","#1B9E77","#E41A1C","pink","#F781BF")
coords<-coords[rownames(SONAR.results),]
#pie plot(large image may need a few minutes)
pie<-SONAR.visualize.pie(SONAR.results,coords,color_vector)
ggsave(pie,filename = paste0(result_path,"pie.pdf"),width = 25,height = 25,units = "in")
#for show
pie
```

### #Spatial scatter plot to show absolute predicted proportion for each cell type 

```{r,warning=FALSE,fig.height=10,fig.width=20}
#cell type proportions show
select_type <- colnames(SONAR.results)[1:8]
##predicted proportion
abs_prop<-SONAR.prop(scale = F,proportion=SONAR.results,spatial_location=coords,select_type=select_type,NumCols = 4)
ggsave(abs_prop,filename = paste0(result_path,"abs_prop.pdf"),width = 25,height = 25,units = "in")
#for show
abs_prop
```

### #Spatial scatter plot to show scaled predicted proportion for each cell type 

```{r,warning=FALSE,fig.height=10,fig.width=20}
##scaled predicted proportion(Min-Max Scales by cell type)
scaled_prop<-SONAR.prop(scale = T,proportion=SONAR.results,spatial_location=coords,select_type=select_type,NumCols = 4,colors = turbo(20,alpha = 0.8))
ggsave(scaled_prop,filename = paste0(result_path,"scaled_prop.pdf"),width = 25,height = 25,units = "in")
#for show
scaled_prop
```

### #Co-localization for pairs of cell types 

```{r,warning=FALSE,fig.height=20,fig.width=20}
##co-localization test
col2 = colorRampPalette(rev(c('#67001F', '#B2182B', '#D6604D', '#F4A582',
                              '#FDDBC7', '#FFFFFF', '#D1E5F0', '#92C5DE',
                              '#4393C3', '#2166AC', '#053061')))
m=par(no.readonly = T)
pdf(paste0(result_path,"colocalization.pdf"),width = 18,height = 18)
par(mfrow=c(1,1))
corrplot(tl.col = "black",pch.col="black",tl.srt = 90,diag = F,order="FPC",col=col2(200),corr =cor(SONAR.results,method="spearman"), p.mat = round(cor_pmat(SONAR.results,method = "spearman"),6),method = "circle",type = "lower",insig="label_sig",sig.level = c(.0001,.001, .01, .05),pch.cex = 1)
dev.off()
par(m)
#for show
corrplot(tl.col = "black",pch.col="black",tl.srt = 90,diag = F,order="FPC",col=col2(200),corr =cor(SONAR.results,method="spearman"), p.mat = round(cor_pmat(SONAR.results,method = "spearman"),6),method = "circle",type = "lower",insig="label_sig",sig.level = c(.0001,.001, .01, .05),pch.cex = 1)
```
